// @temporalio/worker/lib/index.js
"use strict";
/**
 * The temporal worker connects to the service and runs workflows and activities.
 *
 * ### Usage
 *
 * <!--SNIPSTART typescript-hello-worker-->
 * <!--SNIPEND-->
 * @module
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultWorflowInterceptorModules = exports.defaultWorkflowInterceptorModules = exports.bundleWorkflowCode = exports.workflowLogAttributes = exports.WorkflowLogInterceptor = exports.WorkflowInboundLogInterceptor = exports.ReplayError = exports.defaultSinks = exports.appendDefaultInterceptors = exports.Worker = exports.defaultPayloadConverter = exports.CombinedWorkerRunError = exports.makeTelemetryFilterString = exports.Runtime = exports.errors = exports.GracefulShutdownPeriodExpiredError = exports.UnexpectedError = exports.TransportError = exports.ShutdownError = exports.IllegalStateError = exports.startDebugReplayer = exports.NativeConnection = exports.activityLogAttributes = exports.ActivityInboundLogInterceptor = void 0;
var activity_log_interceptor_1 = require("./activity-log-interceptor");
Object.defineProperty(exports, "ActivityInboundLogInterceptor", { enumerable: true, get: function () { return activity_log_interceptor_1.ActivityInboundLogInterceptor; } });
Object.defineProperty(exports, "activityLogAttributes", { enumerable: true, get: function () { return activity_log_interceptor_1.activityLogAttributes; } });
var connection_1 = require("./connection");
Object.defineProperty(exports, "NativeConnection", { enumerable: true, get: function () { return connection_1.NativeConnection; } });
var debug_replayer_1 = require("./debug-replayer");
Object.defineProperty(exports, "startDebugReplayer", { enumerable: true, get: function () { return debug_replayer_1.startDebugReplayer; } });
var common_1 = require("@temporalio/common");
Object.defineProperty(exports, "IllegalStateError", { enumerable: true, get: function () { return common_1.IllegalStateError; } });
var core_bridge_1 = require("@temporalio/core-bridge");
Object.defineProperty(exports, "ShutdownError", { enumerable: true, get: function () { return core_bridge_1.ShutdownError; } });
Object.defineProperty(exports, "TransportError", { enumerable: true, get: function () { return core_bridge_1.TransportError; } });
Object.defineProperty(exports, "UnexpectedError", { enumerable: true, get: function () { return core_bridge_1.UnexpectedError; } });
var errors_1 = require("./errors"); // eslint-disable-line deprecation/deprecation
Object.defineProperty(exports, "GracefulShutdownPeriodExpiredError", { enumerable: true, get: function () { return errors_1.GracefulShutdownPeriodExpiredError; } });
Object.defineProperty(exports, "errors", { enumerable: true, get: function () { return errors_1.errors; } });
__exportStar(require("./interceptors"), exports);
__exportStar(require("./logger"), exports);
var runtime_1 = require("./runtime");
Object.defineProperty(exports, "Runtime", { enumerable: true, get: function () { return runtime_1.Runtime; } });
Object.defineProperty(exports, "makeTelemetryFilterString", { enumerable: true, get: function () { return runtime_1.makeTelemetryFilterString; } });
__exportStar(require("./sinks"), exports);
var worker_1 = require("./worker");
Object.defineProperty(exports, "CombinedWorkerRunError", { enumerable: true, get: function () { return worker_1.CombinedWorkerRunError; } });
Object.defineProperty(exports, "defaultPayloadConverter", { enumerable: true, get: function () { return worker_1.defaultPayloadConverter; } });
Object.defineProperty(exports, "Worker", { enumerable: true, get: function () { return worker_1.Worker; } });
var worker_options_1 = require("./worker-options");
Object.defineProperty(exports, "appendDefaultInterceptors", { enumerable: true, get: function () { return worker_options_1.appendDefaultInterceptors; } });
Object.defineProperty(exports, "defaultSinks", { enumerable: true, get: function () { return worker_options_1.defaultSinks; } });
var replay_1 = require("./replay");
Object.defineProperty(exports, "ReplayError", { enumerable: true, get: function () { return replay_1.ReplayError; } });
var workflow_log_interceptor_1 = require("./workflow-log-interceptor");
Object.defineProperty(exports, "WorkflowInboundLogInterceptor", { enumerable: true, get: function () { return workflow_log_interceptor_1.WorkflowInboundLogInterceptor; } });
Object.defineProperty(exports, "WorkflowLogInterceptor", { enumerable: true, get: function () { return workflow_log_interceptor_1.WorkflowLogInterceptor; } });
Object.defineProperty(exports, "workflowLogAttributes", { enumerable: true, get: function () { return workflow_log_interceptor_1.workflowLogAttributes; } });
var bundler_1 = require("./workflow/bundler");
Object.defineProperty(exports, "bundleWorkflowCode", { enumerable: true, get: function () { return bundler_1.bundleWorkflowCode; } });
Object.defineProperty(exports, "defaultWorkflowInterceptorModules", { enumerable: true, get: function () { return bundler_1.defaultWorkflowInterceptorModules; } });
/**
 * @deprecated Use `defaultWorkflowInterceptorModules` instead
 */
Object.defineProperty(exports, "defaultWorflowInterceptorModules", { enumerable: true, get: function () { return bundler_1.defaultWorkflowInterceptorModules; } });
