// @temporalio/worker/lib/worker-options.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compileWorkerOptions = exports.addDefaultWorkerOptions = exports.appendDefaultInterceptors = exports.defaultSinks = exports.isPathBundleOption = exports.isCodeBundleOption = void 0;
const os = __importStar(require("node:os"));
const v8 = __importStar(require("node:v8"));
const time_1 = require("@temporalio/common/lib/time");
const internal_non_workflow_1 = require("@temporalio/common/lib/internal-non-workflow");
const activity_log_interceptor_1 = require("./activity-log-interceptor");
const runtime_1 = require("./runtime");
const utils_1 = require("./utils");
const bundler_1 = require("./workflow/bundler");
function isCodeBundleOption(bundleOpt) {
    const opt = bundleOpt; // Cast to access properties without TS complaining
    return typeof opt.code === 'string';
}
exports.isCodeBundleOption = isCodeBundleOption;
function isPathBundleOption(bundleOpt) {
    const opt = bundleOpt; // Cast to access properties without TS complaining
    return typeof opt.codePath === 'string';
}
exports.isPathBundleOption = isPathBundleOption;
/**
 * Returns the `defaultWorkerLogger` sink which forwards logs from the Workflow sandbox to a given logger.
 *
 * @param logger a {@link Logger} - defaults to the {@link Runtime} singleton logger.
 */
function defaultSinks(logger) {
    return {
        defaultWorkerLogger: {
            trace: {
                fn(_, message, attrs) {
                    logger ?? (logger = runtime_1.Runtime.instance().logger);
                    logger.trace(message, attrs);
                },
            },
            debug: {
                fn(_, message, attrs) {
                    logger ?? (logger = runtime_1.Runtime.instance().logger);
                    logger.debug(message, attrs);
                },
            },
            info: {
                fn(_, message, attrs) {
                    logger ?? (logger = runtime_1.Runtime.instance().logger);
                    logger.info(message, attrs);
                },
            },
            warn: {
                fn(_, message, attrs) {
                    logger ?? (logger = runtime_1.Runtime.instance().logger);
                    logger.warn(message, attrs);
                },
            },
            error: {
                fn(_, message, attrs) {
                    logger ?? (logger = runtime_1.Runtime.instance().logger);
                    logger.error(message, attrs);
                },
            },
        },
    };
}
exports.defaultSinks = defaultSinks;
/**
 * Appends the default Worker logging interceptors to given interceptor arrays.
 *
 * @param logger a {@link Logger} - defaults to the {@link Runtime} singleton logger.
 */
function appendDefaultInterceptors(interceptors, logger) {
    return {
        activityInbound: [...(interceptors.activityInbound ?? []), (ctx) => new activity_log_interceptor_1.ActivityInboundLogInterceptor(ctx, logger)],
        workflowModules: [...(interceptors.workflowModules ?? []), ...bundler_1.defaultWorkflowInterceptorModules],
    };
}
exports.appendDefaultInterceptors = appendDefaultInterceptors;
function addDefaultWorkerOptions(options) {
    const { buildId, useVersioning, maxCachedWorkflows, showStackTraceSources, namespace, reuseV8Context, sinks, ...rest } = options;
    const debugMode = options.debugMode || isSet(process.env.TEMPORAL_DEBUG);
    const maxConcurrentWorkflowTaskExecutions = options.maxConcurrentWorkflowTaskExecutions ?? 40;
    const maxConcurrentActivityTaskExecutions = options.maxConcurrentActivityTaskExecutions ?? 100;
    const heapSizeMiB = v8.getHeapStatistics().heap_size_limit / utils_1.MiB;
    const defaultMaxCachedWorkflows = reuseV8Context
        ? Math.max(Math.floor((Math.max(heapSizeMiB - 200, 0) * 600) / 1024), 10)
        : Math.max(Math.floor((Math.max(heapSizeMiB - 400, 0) * 250) / 1024), 10);
    if (useVersioning && !buildId) {
        throw new TypeError('Must provide a buildId if useVersioning is true');
    }
    return {
        namespace: namespace ?? 'default',
        identity: `${process.pid}@${os.hostname()}`,
        useVersioning: useVersioning ?? false,
        buildId,
        shutdownGraceTime: 0,
        maxConcurrentLocalActivityExecutions: 100,
        enableNonLocalActivities: true,
        maxConcurrentWorkflowTaskPolls: Math.min(10, maxConcurrentWorkflowTaskExecutions),
        maxConcurrentActivityTaskPolls: Math.min(10, maxConcurrentActivityTaskExecutions),
        stickyQueueScheduleToStartTimeout: '10s',
        maxHeartbeatThrottleInterval: '60s',
        defaultHeartbeatThrottleInterval: '30s',
        // 4294967295ms is the maximum allowed time
        isolateExecutionTimeout: debugMode ? '4294967295ms' : '5s',
        workflowThreadPoolSize: reuseV8Context ? 1 : 2,
        maxCachedWorkflows: maxCachedWorkflows ?? defaultMaxCachedWorkflows,
        enableSDKTracing: false,
        showStackTraceSources: showStackTraceSources ?? false,
        reuseV8Context: reuseV8Context ?? false,
        debugMode: debugMode ?? false,
        interceptors: appendDefaultInterceptors({}),
        sinks: { ...defaultSinks(), ...sinks },
        ...rest,
        maxConcurrentWorkflowTaskExecutions,
        maxConcurrentActivityTaskExecutions,
    };
}
exports.addDefaultWorkerOptions = addDefaultWorkerOptions;
function isSet(env) {
    if (env === undefined)
        return false;
    env = env.toLocaleLowerCase();
    return env === '1' || env === 't' || env === 'true';
}
function compileWorkerOptions(opts) {
    if (opts.maxCachedWorkflows !== 0 && opts.maxCachedWorkflows < 2) {
        runtime_1.Runtime.instance().logger.warn('maxCachedWorkflows must be either 0 (ie. cache is disabled) or greater than 1. Defaulting to 2.');
        opts.maxCachedWorkflows = 2;
    }
    if (opts.maxCachedWorkflows > 0 && opts.maxConcurrentWorkflowTaskExecutions > opts.maxCachedWorkflows) {
        runtime_1.Runtime.instance().logger.warn("maxConcurrentWorkflowTaskExecutions can't exceed maxCachedWorkflows (unless cache is disabled). Defaulting to maxCachedWorkflows.");
        opts.maxConcurrentWorkflowTaskExecutions = opts.maxCachedWorkflows;
    }
    if (opts.maxCachedWorkflows > 0 && opts.maxConcurrentWorkflowTaskExecutions < 2) {
        runtime_1.Runtime.instance().logger.warn("maxConcurrentWorkflowTaskExecutions can't be lower than 2 if maxCachedWorkflows is non-zero. Defaulting to 2.");
        opts.maxConcurrentWorkflowTaskExecutions = 2;
    }
    return {
        ...opts,
        shutdownGraceTimeMs: (0, time_1.msToNumber)(opts.shutdownGraceTime),
        shutdownForceTimeMs: (0, time_1.msOptionalToNumber)(opts.shutdownForceTime),
        stickyQueueScheduleToStartTimeoutMs: (0, time_1.msToNumber)(opts.stickyQueueScheduleToStartTimeout),
        isolateExecutionTimeoutMs: (0, time_1.msToNumber)(opts.isolateExecutionTimeout),
        maxHeartbeatThrottleIntervalMs: (0, time_1.msToNumber)(opts.maxHeartbeatThrottleInterval),
        defaultHeartbeatThrottleIntervalMs: (0, time_1.msToNumber)(opts.defaultHeartbeatThrottleInterval),
        loadedDataConverter: (0, internal_non_workflow_1.loadDataConverter)(opts.dataConverter),
    };
}
exports.compileWorkerOptions = compileWorkerOptions;
